File: src/main/java/org/warp/picalculator/math/rules/VariableRule2.java
package org.warp.picalculator.math.rules;

import org.warp.picalculator.Error;
import org.warp.picalculator.math.Function;
import org.warp.picalculator.math.FunctionOperator;
import org.warp.picalculator.math.MathContext;
import org.warp.picalculator.math.functions.Expression;
import org.warp.picalculator.math.functions.Multiplication;
import org.warp.picalculator.math.functions.Number;
import org.warp.picalculator.math.functions.Subtraction;
import org.warp.picalculator.math.functions.Sum;

import it.unimi.dsi.fastutil.objects.ObjectArrayList;

/**
 * Variable rule<br>
 * <b>ax+x=(a+1)*x (a,b NUMBER; x VARIABLES)</b>
 * 
 * @author Andrea Cavalli
 *
 */
public class VariableRule2 {

	public static boolean compare(FunctionOperator fnc) {
		if (fnc.getParameter1() instanceof Multiplication) {
			final Multiplication m1 = (Multiplication) fnc.getParameter1();
			if (m1.getParameter2().equals(fnc.getParameter2())) {
				return true;
			}
		}
		return false;
	}

	public static ObjectArrayList<Function> execute(FunctionOperator fnc) throws Error {
		final MathContext root = fnc.getMathContext();
		final ObjectArrayList<Function> result = new ObjectArrayList<>();
		final Multiplication m1 = (Multiplication) fnc.getParameter1();
		final Function a = m1.getParameter1();
		final Function x = fnc.getParameter2();

		FunctionOperator rets;
		if (fnc instanceof Sum) {
			rets = new Sum(root, a, new Number(root, 1));
		} else {
			rets = new Subtraction(root, a, new Number(root, 1));
		}
		final Expression rete = new Expression(root, rets);
		final Multiplication retm = new Multiplication(root, rete, x);
		result.add(retm);
		return result;
	}

}

File: src/main/resources/rules/VariableRule2.js
//Imports


var Error = org.warp.picalculator.Error;
var Function = org.warp.picalculator.math.Function;
var FunctionOperator = org.warp.picalculator.math.FunctionOperator;
var MathContext = org.warp.picalculator.math.MathContext;
var Multiplication = org.warp.picalculator.math.functions.Multiplication;
var Number = org.warp.picalculator.math.functions.Number;
var Subtraction = org.warp.picalculator.math.functions.Subtraction;
var Sum = org.warp.picalculator.math.functions.Sum;

var ObjectArrayList = Java.type("it.unimi.dsi.fastutil.objects.ObjectArrayList");

/**
 * Variable rule
 * ax+x=(a+1)*x (a,b NUMBER; x VARIABLES)
 * 
 * @author Andrea Cavalli
 *
 */
var rule = {
	// Rule name
	getRuleName: function() {
		return "VariableRule2";
	},
	// Rule type
	getRuleType: function() {
		return RuleType.REDUCTION;
	},
	/* Rule function
	   Returns:
	     - null if it's not executable on the function "f"
	     - An ObjectArrayList<Function> if it did something
	*/

	execute: function(f) {
		var isExecutable;
		var fnc = f;
		if (ScriptUtils.instanceOf(f, Sum.class) || ScriptUtils.instanceOf(f, Subtraction.class)) {
			if (ScriptUtils.instanceOf(fnc.getParameter1(), Multiplication.class)) {
				var m1 = fnc.getParameter1();
				if (m1.getParameter2().equals(fnc.getParameter2())) {
					isExecutable = true;
				}
			}
		}
		
		if (isExecutable) {
			var root = fnc.getMathContext();
			var result = new ObjectArrayList();
			var m1 = fnc.getParameter1();
			var a = m1.getParameter1();
			var x = fnc.getParameter2();
	
			var rets;
			if (ScriptUtils.instanceOf(fnc, Sum.class)) {
				rets = new Sum(root, a, new Number(root, 1));
			} else {
				rets = new Subtraction(root, a, new Number(root, 1));
			}
			var retm = new Multiplication(root, rets, x);
			result.add(retm);
			return result;
		} else {
			return null;
		}
	}
}

//Add this rule to the list of rules
RulesManager.addRule(engine.getInterface(rule, Rule.class));